### PR TITLE
Create a specific calculateRootHash method for BonsaiInMemoryWorldState to improve performance

### PR DESCRIPTION
**Pull request by ahamlat**
## PR description
Create a specific calculateRootHash for BonsaiInMemoryWorldState, avoid committing and rollbacking as it is not necessary. This PR parallelize also the accounts update storage in calculateRootHash method.

The yellow line below is the version with the modifications in this PR, the other node is used as a reference node (a baseline) regarding block size and the number of transactions per block. 

image

We have below engine_newPayloadV1 RPC call (Block processing) mean time comparaison between this PR, version 22.7.6 and version 22.7.7

image

We can notice also that calculateRootHash method is consuming less (real) time on CPU because of the parallelization and the fact we removed commit and rollback method.

**Before this PR**
image

**After this PR**
image

## Fixed Issue(s)

Mitigate #4549

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

**Comment by macfarla** [comment]
These files required manual handling in commit [da8904d](https://github.com/hyperledger/besu/pull/4568/commits/da8904dfbf3486717f601bc88ed699fffa6b0224):
    both modified:   ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/BlockchainQueries.java
    both modified:   ethereum/blockcreation/src/main/java/org/hyperledger/besu/ethereum/blockcreation/AbstractBlockCreator.java
    both modified:   ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiPersistedWorldState.java
    both modified:   ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiSnapshotWorldState.java
    both modified:   ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateArchive.java
    both modified:   ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateUpdater.java
    both modified:   ethereum/core/src/test/java/org/hyperledger/besu/ethereum/MainnetBlockValidatorTest.java
    both added:      ethereum/core/src/test/java/org/hyperledger/besu/ethereum/bonsai/BonsaiSnapshotWorldStateArchiveTest.java
    both modified:   plugin-api/build.gradle

**Comment by garyschulte** [comment]
@ahamlat - I took the liberty of rebasing this off of main.  the merges of main and rc3 branch were making the commit history too hard to follow.

**Review by macfarla — APPROVED** [review]
LGTM

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiInMemoryWorldState.java [hunk 1]
```diff
@@ -38,12 +46,90 @@ public Hash rootHash() {
...
 
   public Hash rootHash(final BonsaiWorldStateUpdater localUpdater) {
-    final BonsaiWorldStateKeyValueStorage.BonsaiUpdater updater = worldStateStorage.updater();
-    try {
-      final Hash calculatedRootHash = calculateRootHash(updater, localUpdater);
-      return Hash.wrap(calculatedRootHash);
-    } finally {
-      updater.rollback();
+    final Hash calculatedRootHash = calculateRootHash(localUpdater);
+    return Hash.wrap(calculatedRootHash);
```
Comment (matkt) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiInMemoryWorldState.java:50]: it can be one line

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiPersistedWorldState.java [hunk 2]
```diff
@@ -104,43 +104,76 @@ public BonsaiWorldStateKeyValueStorage getWorldStateStorage() {
...
-              .orElse(null);
-      if (oldAccount == null) {
-        // This is when an account is both created and deleted within the scope of the same
-        // block.  A not-uncommon DeFi bot pattern.
-        continue;
+    clearStorage(stateUpdater, worldStateUpdater);
+
+    // This must be done before updating the accounts so
+    // that we can get the storage state hash
+    updateAccountStorageState(stateUpdater, worldStateUpdater);
```
Comment (matkt) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiPersistedWorldState.java:111]: In a next PR it will be nice to add also this modification here https://github.com/hyperledger/besu/pull/4568/files#diff-0cb922a2c0562efbff99afb7bbde477d58a8adb52a82b076756ac0d90e387fe6R58

Comment (ahamlat) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiPersistedWorldState.java:111]: I'm going to make a next PR to apply the same modifications to BonsaiPersistedWorldState.
